### PR TITLE
ELPP-3429 Fix race condition on profile creation

### PR DESCRIPTION
**Pull request by giorgiosironi**
Assumes `profiles` points to `orcid-dummy`.

1. Clear the database
2. Perform two /token requests concurrently, which create the same profile

Seems easy to reproduce:
```
{
  "name": "profiles.api.errors",
  "message": "(raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely) (psycopg2.IntegrityError) duplicate key value violates unique constraint \"profile_orcid_key\"\nDETAIL:  Key (orcid)=(0000-0002-1825-0097) already exists.\n [SQL: 'INSERT INTO profile (id, preferred_name, index_name, orcid) VALUES (%(id)s, %(preferred_name)s, %(index_name)s, %(orcid)s)'] [parameters: {'index_name': 'Carberry, Josiah', 'id': 'muls9j78', 'orcid': '0000-0002-1825-0097', 'preferred_name': 'Josiah Carberry'}]",
  "asctime": "2018-02-02 09:17:43,893",
  "levelname": "ERROR",
  "exc_info": {
    "trace": [
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/flask/app.py\", line 1612, in full_dispatch_request\n    rv = self.dispatch_request()\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/flask/app.py\", line 1598, in dispatch_request\n    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"./profiles/utilities.py\", line 53, in wrapper\n    response = func(*args, **kwargs)\n",
      "  File \"./profiles/api/oauth2.py\", line 139, in _token\n    _find_and_update_access_token(json_data)\n",
      "  File \"./profiles/api/oauth2.py\", line 170, in _find_and_update_access_token\n    orcid_token = orcid_tokens.get(token_data['orcid'])\n",
      "  File \"./profiles/repositories.py\", line 65, in get\n    return self.db.session.query(OrcidToken).filter_by(orcid=orcid).one()\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/query.py\", line 2814, in one\n    ret = self.one_or_none()\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/query.py\", line 2784, in one_or_none\n    ret = list(self)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/query.py\", line 2854, in __iter__\n    self.session._autoflush()\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/session.py\", line 1407, in _autoflush\n    util.raise_from_cause(e)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/util/compat.py\", line 203, in raise_from_cause\n    reraise(type(exception), exception, tb=exc_tb, cause=cause)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/util/compat.py\", line 187, in reraise\n    raise value\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/session.py\", line 1397, in _autoflush\n    self.flush()\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/session.py\", line 2177, in flush\n    self._flush(objects)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/session.py\", line 2297, in _flush\n    transaction.rollback(_capture_exception=True)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/util/langhelpers.py\", line 66, in __exit__\n    compat.reraise(exc_type, exc_value, exc_tb)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/util/compat.py\", line 187, in reraise\n    raise value\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/session.py\", line 2261, in _flush\n    flush_context.execute()\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/unitofwork.py\", line 389, in execute\n    rec.execute(self)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/unitofwork.py\", line 548, in execute\n    uow\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/persistence.py\", line 181, in save_obj\n    mapper, table, insert)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/orm/persistence.py\", line 799, in _emit_insert_statements\n    execute(statement, multiparams)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/engine/base.py\", line 945, in execute\n    return meth(self, multiparams, params)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/sql/elements.py\", line 263, in _execute_on_connection\n    return connection._execute_clauseelement(self, multiparams, params)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/engine/base.py\", line 1053, in _execute_clauseelement\n    compiled_sql, distilled_params\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/engine/base.py\", line 1189, in _execute_context\n    context)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/engine/base.py\", line 1402, in _handle_dbapi_exception\n    exc_info\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/util/compat.py\", line 203, in raise_from_cause\n    reraise(type(exception), exception, tb=exc_tb, cause=cause)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/util/compat.py\", line 186, in reraise\n    raise value.with_traceback(tb)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/engine/base.py\", line 1182, in _execute_context\n    context)\n",
      "  File \"/srv/profiles/venv/lib/python3.5/site-packages/sqlalchemy/engine/default.py\", line 470, in do_execute\n    cursor.execute(statement, parameters)\n"
    ],
    "message": "Exception: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely) (psycopg2.IntegrityError) duplicate key value violates unique constraint \"profile_orcid_key\"\nDETAIL:  Key (orcid)=(0000-0002-1825-0097) already exists.\n [SQL: 'INSERT INTO profile (id, preferred_name, index_name, orcid) VALUES (%(id)s, %(preferred_name)s, %(index_name)s, %(orcid)s)'] [parameters: {'index_name': 'Carberry, Josiah', 'id': 'muls9j78', 'orcid': '0000-0002-1825-0097', 'preferred_name': 'Josiah Carberry'}]",
    "class": "IntegrityError"
  }
}
```

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
@giorgiosironi Could you update this to use the containers? Had a quick go but doesn't seem to show a failure, just hangs.

**Comment by thewilkybarkid** [comment]
@giorgiosironi Fixed, and now tested. Not deleted the scripts yet.

File: profiles/repositories.py [hunk 1]
```diff
@@ -89,8 +93,22 @@ def generate_id():
...
+    def add(self, profile: Profile) -> Profile:
+        self.db.session.begin_nested()
         self.db.session.add(profile)
+        try:
+            self.db.session.commit()
+        except IntegrityError as exception:
+            self.db.session.rollback()
+
+            if profile.orcid:
+                LOGGER.info('Profile for ORCID %s appears to already exist', profile.orcid)
```
Comment (giorgiosironi) [inline profiles/repositories.py]: assumes ORCID is the only field that can cause an `IntegrityError`? A more conservative message could be `Cannot insert profile, assuming its existence to retrieve it` or similar

File: test/conftest.py [hunk 2]
```diff
@@ -83,6 +85,17 @@ def database(app: Flask, request: FixtureRequest) -> SQLAlchemy:
     db.app = app
     db.create_all()
 
+    # pylint:disable=unused-argument
+    # pylint:disable=unused-variable
+    @event.listens_for(db.engine, 'connect')
+    def do_connect(connection: Connection, *args) -> None:
```
Comment (giorgiosironi) [inline test/conftest.py]: some comments on why these two hooks are necessary in tests? Related to SQLite?

File: profiles/repositories.py (hunk 1 above)
Comment (thewilkybarkid) [reply profiles/repositories.py]: Could be email address too. Will expand.

**Comment by giorgiosironi** [comment]
Can't push `Approve` because I opened this, but :+1:

**Pull request merged**